src/Microsoft.OpenApi/Models/OpenApiModelFactory.cs

```diff
@@ -0,0 +1,221 @@
+        /// </summary>
+        /// <param name="input">The input string.</param>
+        /// <param name="version"></param>
+        /// <param name="diagnostic">The diagnostic entity containing information from the reading process.</param>
+        /// <param name="format">The Open API format</param>
+        /// <param name="settings">The OpenApi reader settings.</param>
+        /// <returns>An OpenAPI document instance.</returns>
+        public static T Parse<T>(string input,
+                                 OpenApiSpecVersion version,
+                                 out OpenApiDiagnostic diagnostic,
+                                 string format = null,
+                                 OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        {
+            format ??= OpenApiConstants.Json;
+            return OpenApiReaderRegistry.GetReader(format).Parse<T>(input, version, out diagnostic, settings);
+        }
+
+        public static T Load<T>(string url, OpenApiSpecVersion version, out OpenApiDiagnostic diagnostic, OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        {
+            var format = GetFormat(url);
+            return OpenApiReaderRegistry.GetReader(format).Read<T>(url, version, out diagnostic, settings);
+        }
+
+        public static T Load<T>(Stream input, OpenApiSpecVersion version, out OpenApiDiagnostic diagnostic, string format, OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        {
+            format ??= OpenApiConstants.Json;
+            return OpenApiReaderRegistry.GetReader(format).Read<T>(input, version, out diagnostic, settings);
+        }
+
+        public static T Load<T>(TextReader input, OpenApiSpecVersion version, out OpenApiDiagnostic diagnostic, string format, OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        {
+            format ??= OpenApiConstants.Json;
+            return OpenApiReaderRegistry.GetReader(format).Read<T>(input, version, out diagnostic, settings);
+        }
+
+
+        private static string GetContentType(string url)
+        {
+            var response = _httpClient.GetAsync(url).GetAwaiter().GetResult();
+            var contentType = response.Content.Headers.ContentType.MediaType;
+            if (contentType.EndsWith(OpenApiConstants.Json, StringComparison.OrdinalIgnoreCase))
+            {
+                return OpenApiConstants.Json;
+            }
+            else if (contentType.EndsWith(OpenApiConstants.Yaml, StringComparison.OrdinalIgnoreCase))
+            {
+                return OpenApiConstants.Yaml;
+            }
+            return null;
+        }
+
+        private static string GetFormat(string url)
+        {
+            if (!string.IsNullOrEmpty(url))
+            {
+                if (url.StartsWith("http", StringComparison.OrdinalIgnoreCase) || url.StartsWith("https", StringComparison.OrdinalIgnoreCase))
```

Comment:
Out of curiosity are yaml/json the only formats that could exist? If not, it may make sense to simplify this to only pull the suffix and use for lookup in the registry(in the event that the caller already registered that format in the registry)

Reply:
Yeah using the suffix for a registry lookup sounds more effective. Let me update this